modules/misc/listenbrainz.c

```diff
@@ -0,0 +1,585 @@
+    {
+        ReadMetaData (intf);
+        return;
+    }
+
+    if ( state == VLC_PLAYER_STATE_STOPPED )
+        Enqueue (intf);
+}
+
+static void OnTimerUpdate (const struct vlc_player_timer_point *value, void *data)
+{
+    intf_thread_t *intf = data;
+    intf_sys_t *p_sys = intf->p_sys;
+    p_sys->time_played = SEC_FROM_VLC_TICK (value->ts - VLC_TICK_0);
+}
+
+static void OnTimerStopped (vlc_tick_t system_date, void *data){}
+
+static void PlaylistItemChanged (vlc_playlist_t *playlist, ssize_t index, void *data)
+{
+    VLC_UNUSED (index);
+
+    intf_thread_t *intf = data;
+    if ( index > 0 )
+        Enqueue (intf);
+
+    intf_sys_t *p_sys = intf->p_sys;
+    p_sys->b_meta_read = false;
+
+    vlc_player_t *player = vlc_playlist_GetPlayer (playlist);
+    input_item_t *item = vlc_player_GetCurrentMedia (player);
+
+    if ( !item || vlc_player_GetVideoTrackCount (player) )
+    {
+        msg_Dbg (intf, "Invalid item or not an audio-only input.");
+        return;
+    }
+
+    p_sys->time_played = 0;
+
+    if ( input_item_IsPreparsed (item) )
+        ReadMetaData (intf);
+}
+
+static char* PreparePayload (intf_thread_t *p_this)
+{
+    intf_sys_t *p_sys = p_this->p_sys;
+    struct vlc_memstream payload;
+    vlc_memstream_open (&payload);
+
+    vlc_mutex_lock (&p_sys->lock);
+
+    if ( p_sys->queue.size == 1 )
+        vlc_memstream_printf (&payload, "{\"listen_type\":\"single\",\"payload\":[");
+    else
+        vlc_memstream_printf (&payload, "{\"listen_type\":\"import\",\"payload\":[");
+
+    for ( int i_song = 0; i_song < (int) p_sys->queue.size; i_song++ )
+    {
+        listen_t *p_song = &p_sys->queue.data[i_song];
+
+        vlc_memstream_printf (&payload, "{\"listened_at\": %"PRIu64, (uint64_t) p_song->date);
+        vlc_memstream_printf (&payload, ", \"track_metadata\": {\"artist_name\": \"%s\", ",
+                              vlc_uri_decode (p_song->psz_artist));
+        vlc_memstream_printf (&payload, " \"track_name\": \"%s\", ", vlc_uri_decode (p_song->psz_title));
+        if ( !EMPTY_STR (p_song->psz_album) )
+            vlc_memstream_printf (&payload, " \"release_name\": \"%s\"", vlc_uri_decode (p_song->psz_album));
+        if ( !EMPTY_STR (p_song->psz_musicbrainz_id) )
+            vlc_memstream_printf (&payload, ", \"additional_info\": {\"recording_mbid\":\"%s\"} ",
+                                  vlc_uri_decode (p_song->psz_musicbrainz_id));
+        vlc_memstream_printf (&payload, "}}");
+    }
+
+    vlc_memstream_printf (&payload, "]}");
+    vlc_mutex_unlock (&p_sys->lock);
+
+    int i_status = vlc_memstream_close (&payload);
+    if ( !i_status )
+    {
+        msg_Dbg (p_this, "Payload: %s", payload.ptr);
+        return payload.ptr;
+    }
+    else
+        return NULL;
+}
+
+static char* PrepareRequest (intf_thread_t *p_this, char* payload)
+{
+    intf_sys_t *p_sys = p_this->p_sys;
+    struct vlc_memstream request;
+
+    vlc_mutex_lock (&p_sys->lock);
```

Comment:
The lock here is useless.
Indeed, p_sys->p_submit_url.psz_path, p_sys->p_submit_url.psz_host and p_sys->p_submit_url.psz_host are set in the Open function before the thread is created.